### PR TITLE
Update the `reportportal` plugin

### PR DESCRIPTION
**Pull request by 4N0body5**
Features:
* support of environment variables `TMT_PLUGIN_REPORT_REPORTPORTAL_<option>` also for report step set via fmf
* launch-description option modified to append an existing description instead of overwriting it
* option to specify api-version

Pull Request Checklist

* [x] implement the feature 
* [x] write the documentation
* [x] modify the json schema

Plan for additional PR:
* rerun of failed report for older ReportPortal version
* test coverage

### CODE REVIEW

File: tmt/steps/report/reportportal.py [hunk 1]
```diff
@@ -36,50 +36,51 @@ class ReportReportPortalData(tmt.steps.report.ReportStepData):
...
-             "option (Otherwise Summary from plan fmf data per each launch is used by default).")
+        metavar="DESCRIPTION",
+        default=os.getenv('TMT_PLUGIN_REPORT_REPORTPORTAL_DESCRIPTION'),
+        help="Pass the description for ReportPortal launch with '--suite-per-plan' option"
+             "or append the original (plan summary) with additional info."
+             "Appends suite/test description with upload-to-launch/suite options.")
     launch_per_plan: bool = field(
         option="--launch-per-plan",
-        default=False,
+        default=bool(os.getenv('TMT_PLUGIN_REPORT_REPORTPORTAL_LAUNCH_PER_PLAN')),
```
Comment (happz) [inline tmt/steps/report/reportportal.py:46]: `0` would be translated to `True`, which would be 1. puzzling, 2. incorrect. It's a non-empty string, therefore it's true-ish:

```
happz@multivac ~ $ export TMT_PLUGIN_REPORT_REPORTPORTAL_LAUNCH_PER_PLAN="0"
happz@multivac ~ $ python
Python 3.11.8 (main, Feb  7 2024, 00:00:00) [GCC 13.2.1 20231011 (Red Hat 13.2.1-4)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> bool(os.getenv('TMT_PLUGIN_REPORT_REPORTPORTAL_LAUNCH_PER_PLAN'))
True
>>> 
```

This should work:

```python
bool(os.getenv('TMT_PLUGIN_REPORT_REPORTPORTAL_LAUNCH_PER_PLAN') == '1')
```

File: tmt/steps/report/reportportal.py [hunk 2]
```diff
@@ -102,12 +108,10 @@ class ReportReportPortal(tmt.steps.report.ReportPlugin[ReportReportPortalData]):
...
 
-    * token for authentication
+    * TOKEN for authentication
     * URL of the ReportPortal instance
-    * project name
-    * optional API version to override the default one (v1)
-    * optional launch name to override the deafult name based on the tmt
-      plan name
+    * PROJECT name
+    * [LAUNCH] name to override the default name based on the tmt plan name
```
Comment (happz) [inline tmt/steps/report/reportportal.py:114]: Please, don't undo these changes. I tried to improve the docstring, aligning the plugin more with the rest of the flock.

For example, there is no such thing as "PROJECT" - there is `--project` option, there is a `project` key, and both refer to the `PROJECT_NAME` metavar. There's no "PROJECT", in capital letters. On the other hand, one should provide a "project name" - either by option or a key, hence it's not referring to one particular way on purpose - token, and optionally "launch name" or "API version".

`[LAUNCH]` is again very hard to follow: there are `--launch` and `launch` and `LAUNCH_NAME`, but no `LAUNCH` in the help texts of plugin keys/options. Plus its optionality is already expressed in the text - `[]` is commonly used with CLI options, but may not be that clear for someone editing an fmf file.

File: tmt/steps/report/reportportal.py [hunk 3]
```diff
@@ -147,8 +149,13 @@ class ReportReportPortal(tmt.steps.report.ReportPlugin[ReportReportPortalData]):
...
       resulting in one launch only, and one or more suites within.
-      (Recommended to define launch-name and launch-description in
-      addition)
+      (Recommended to define launch-name and launch-description in addition)
+
+    To override the default reportportal API version (v1):
+
+    .. code-block:: bash
+
+        export TMT_PLUGIN_REPORT_REPORTPORTAL_API_VERSION="v2"
```
Comment (happz) [inline tmt/steps/report/reportportal.py:158]: Is this bit necessary? It focuses on CLI only without telling it to the user. I may easily override RP API in fmf file as well, right? I can easily override many other keys too, yet they are not mentioned. Written like this, it seems like an ad-hoc addition, and I for one would remove it, because on line 116 I already see a paragraph describing how one can override keys on the command line.

File: tmt/steps/report/reportportal.py [hunk 4]
```diff
@@ -116,10 +120,8 @@ class ReportReportPortal(tmt.steps.report.ReportPlugin[ReportReportPortalData]):
 
         export TMT_PLUGIN_REPORT_REPORTPORTAL_${MY_OPTION}=${MY_VALUE}
 
-    The optional launch name doesn't have to be provided if it is the
-    same as the plan name (by default). Assuming the URL and token
-    are provided by the environment variables, the plan config can look
-    like this:
+    Assuming the URL and token are provided by the environment variables,
+    the plan config can look like this:
```
Comment (happz) [inline tmt/steps/report/reportportal.py:124]: Is the statement about the launch name no longer true? It seems to be still correct, judging by line 313, `launch_name = self.data.launch or self.step.plan.name`.

File: tmt/steps/report/reportportal.py (hunk 1 above)
Comment (4N0body5) [reply tmt/steps/report/reportportal.py:46]: You're right. Thanks a lot.

File: tmt/steps/report/reportportal.py (hunk 4 above)
Comment (4N0body5) [reply tmt/steps/report/reportportal.py:124]: It is true, but I find it unnecessary there, if it's just a few lines above mentioned as optional. And it is also mentioned in options' description, see `launch` and `suite-per-plan`. The original sentence could be confusing, as it uses plan summary of the first plan only with option --suite-per-plan.

File: tmt/steps/report/reportportal.py (hunk 1 above)
Comment (4N0body5) [reply tmt/steps/report/reportportal.py:46]: What do you think of this?
bool(os.getenv('TMT_PLUGIN_REPORT_REPORTPORTAL_LAUNCH_PER_PLAN') in ('True', 'true', '1'))

File: tmt/steps/report/reportportal.py [hunk 5]
```diff
@@ -14,85 +14,124 @@
...
         help="Name of the project into which the results should be uploaded.")
+
     launch: Optional[str] = field(
         option="--launch",
         metavar="LAUNCH_NAME",
-        default=os.getenv('TMT_PLUGIN_REPORT_REPORTPORTAL_LAUNCH'),
-        help="The launch name (name of plan per launch is used by default).")
+        default=_str_env_to_default('launch', None),
+        help="Set the launch name, otherwise name of the plan is used by default. Should be"
+             " defined with suite-per-plan option or it will be named after the first plan")
```
Comment (happz) [inline tmt/steps/report/reportportal.py:55]: `help` can be a multi-line string, which is better than multiple joined strings (no easy-to-forget leading/trailing spaces):

```python
        help="""
             Set the launch name, otherwise name of the plan is used by default.
             Should be defined with suite-per-plan option or it will be named
             after the first plan
             """)
```

File: tmt/steps/report/reportportal.py [hunk 6]
```diff
@@ -14,85 +14,124 @@
...
+        default=str(_str_env_to_default('defect_type', "^TMT_.*")),
         help="Regular expression for excluding environment variables "
              "from reporting to ReportPortal ('^TMT_.*' used by default)."
              "Parameters in ReportPortal get filtered out by the pattern"
              "to prevent overloading and to preserve the history aggregation"
              "for ReportPortal item if tmt id is not provided")
 
+    api_version: str = field(
+        option="--api-version",
+        metavar="PATTERN",
```
Comment (happz) [inline tmt/steps/report/reportportal.py:119]: Pattern? That suggests regular expression. Although it looks more like a copy-paste relict...

File: tmt/steps/report/reportportal.py [hunk 7]
```diff
@@ -14,85 +14,124 @@
...
-        default=None,
+        default=_str_env_to_default('defect_type', None),
         help="Pass the defect type to be used for failed test, which is defined in the project"
              " (e.g. 'Idle'). 'To Investigate' is used by default.")
+
     exclude_variables: str = field(
         option="--exclude-variables",
         metavar="PATTERN",
-        default="^TMT_.*",
+        default=str(_str_env_to_default('defect_type', "^TMT_.*")),
```
Comment (happz) [inline tmt/steps/report/reportportal.py:110]: `str()` call is superfluous.

Comment (happz) [reply tmt/steps/report/reportportal.py:110]: Aha, it's probably not, it turns `None` into `"None"`. Can the type of `exclude_variables` be `Optional[str]`?

Comment (happz) [reply tmt/steps/report/reportportal.py:110]: Nevermind, that would be wrong as there's a defined default value...

File: tmt/steps/report/reportportal.py [hunk 8]
```diff
@@ -203,8 +234,7 @@ def get_headers(self) -> dict[str, str]:
                 "Content-Type": "application/json"}
 
     def get_url(self) -> str:
-        api_version = os.getenv(
-            'TMT_PLUGIN_REPORT_REPORTPORTAL_API_VERSION') or self.DEFAULT_API_VERSION
+        api_version = self.data.api_version
```
Comment (happz) [inline tmt/steps/report/reportportal.py:237]: You can inject `self.data.version` into the string directly, is there a need for `api_version` variable?

File: tmt/steps/report/reportportal.py [hunk 9]
```diff
@@ -14,85 +14,124 @@
...
+
+    @staticmethod
+    def _flag_env_to_default(option: str, default: bool) -> bool:
+        env_var = 'TMT_PLUGIN_REPORT_REPORTPORTAL_' + option.upper()
+        if env_var not in os.environ:
+            return default
+        return bool(os.getenv(env_var) == '1')
+
+    @staticmethod
+    def _str_env_to_default(option: str, default: Optional[str]) -> Optional[str]:
```
Comment (happz) [inline tmt/steps/report/reportportal.py:26]: So, I think it's time for `@overload`. Because you have two methods with different return value types: if `default` is an actual string, `_str_env_to_default()` will always return a string, never `None` (either envvar exists, or default is returned). And it's possible to distinguish them on the typing level, with `@overload` defining variants:

```python
@overload
@staticmethod
def _str_env_to_default(option: str, default: None) -> Optional[str]:
    pass

@overload
@staticmethod
def _str_env_to_default(option: str, default: str) -> str:
    pass

@staticmethod
def _str_env_to_default(option: str, default: Optional[str]) -> str:
    env_var = 'TMT_PLUGIN_REPORT_REPORTPORTAL_' + option.upper()
    if env_var not in os.environ or os.getenv(env_var) is None:
        return default
    return str(os.getenv(env_var))
```

This should remove the need for `str()` calls below, and mypy should be able to realize that `_str_env_to_default('defect_type', "^TMT_.*")` returns `str` which matches the declared type of `exclude_variables`.

File: tmt/steps/report/reportportal.py (hunk 7 above)
Comment (happz) [reply tmt/steps/report/reportportal.py:110]: See https://github.com/teemtee/tmt/pull/2702#discussion_r1503818742

Comment (4N0body5) [reply tmt/steps/report/reportportal.py:110]: It should never be None in this case. So can I leave it like this, or better change it to Optional[str]?

File: tmt/steps/report/reportportal.py (hunk 6 above)
Comment (4N0body5) [reply tmt/steps/report/reportportal.py:119]: Thanks, it's copy-paste mistake

File: tmt/steps/report/reportportal.py [hunk 10]
```diff
@@ -14,77 +14,139 @@
...
+
+    @staticmethod
+    def _flag_env_to_default(option: str, default: bool) -> bool:
+        env_var = 'TMT_PLUGIN_REPORT_REPORTPORTAL_' + option.upper()
+        if env_var not in os.environ:
+            return default
+        return bool(os.getenv(env_var) == '1')
+
+    @overload
+    @staticmethod
```
Comment (happz) [inline tmt/steps/report/reportportal.py:26]: I guess the combination of `staticmethod` and `overload` is confusing package build on RHEL9, ruining the tests. I'm afraid these methods would need to become functions, i.e. be moved outside of the `ReportReportPortalData` class.

**Comment by psss** [comment]
This is an important blocker for compose qualification testing, proposing for `1.31`.

File: tmt/steps/report/reportportal.py [hunk 11]
```diff
@@ -12,79 +12,141 @@
...
-             "Parameters in ReportPortal get filtered out by the pattern"
-             "to prevent overloading and to preserve the history aggregation"
-             "for ReportPortal item if tmt id is not provided")
+        default=_str_env_to_default('defect_type', "^TMT_.*"),
+        help="""
+             Regular expression for excluding environment variables
+             from reporting to ReportPortal ('^TMT_.*' used by default).
+             Parameters in ReportPortal get filtered out by the pattern
+             to prevent overloading and to preserve the history aggregation
+             for ReportPortal item if tmt id is not provided
```
Comment (psss) [inline tmt/steps/report/reportportal.py:142]:
```suggestion
             for ReportPortal item if tmt id is not provided.
```

File: tmt/steps/report/reportportal.py [hunk 12]
```diff
@@ -12,79 +12,141 @@
...
+             """)
+
+    api_version: str = field(
+        option="--api-version",
+        metavar="VERSION",
+        default=_str_env_to_default('api_version', "v1"),
+        help="Override the default reportportal API version (v1)")
```
Comment (psss) [inline tmt/steps/report/reportportal.py:149]:
```suggestion
        help="Override the default reportportal API version (v1).")
```

File: tmt/steps/report/reportportal.py [hunk 13]
```diff
@@ -248,6 +300,14 @@ def get_rp_api(self, session: requests.Session, data_path: str) -> requests.Resp
         self.handle_response(response)
         return response
 
+    def append_description(self, curr_description: str) -> str:
```
Comment (psss) [inline tmt/steps/report/reportportal.py:303]: Please add a docstring.

**Review by psss — CHANGES_REQUESTED** [review]
Thanks for improving this, looks good. Added just a few minor comments.

File: tmt/steps/report/reportportal.py [hunk 14]
```diff
@@ -12,79 +12,141 @@
...
 
+def _flag_env_to_default(option: str, default: bool) -> bool:
+    env_var = 'TMT_PLUGIN_REPORT_REPORTPORTAL_' + option.upper()
+    if env_var not in os.environ:
+        return default
+    return bool(os.getenv(env_var) == '1')
+
+
+@overload
+def _str_env_to_default(option: str, default: None) -> Optional[str]:
```
[the remark this answers is not in the document]
Comment (psss) [reply tmt/steps/report/reportportal.py:23]: happz explained this to me, sorry for the noise.

File: tmt/steps/report/reportportal.py (hunk 11 above)
Comment (psss) [reply tmt/steps/report/reportportal.py]: Fixed in aab14887.

File: tmt/steps/report/reportportal.py (hunk 12 above)
Comment (psss) [reply tmt/steps/report/reportportal.py]: Fixed in aab14887.

File: tmt/steps/report/reportportal.py (hunk 13 above)
Comment (psss) [reply tmt/steps/report/reportportal.py:303]: Fixed in aab14887.

**Review by psss — APPROVED** [review]
Thanks for improving this!
